### ISSUE TITLE
Implement AdamW (and AdamWR) optimizer

### ISSUE THREAD

**Issue by stolsvik** [deeplearning4j/deeplearning4j#5843]
According to this blogpost: http://www.fast.ai/2018/07/02/adam-weight-decay/ and mentioned article https://arxiv.org/abs/1711.05101, Adam has problems when used with L2 regularization. If i understand correctly, the problem stems from the fact that L2 is equivalent with Weight Decay when used with SGD, but it is not equivalent when used with adaptive optimizers. "AdamW" fixes this, with the weight decay then being set directly on the optimizer. AdamWR is apparently another iteration with "warm restarts" that generates state of the art results on some ML problems.

**Comment by tom-adsfund** [comment]
(My opinions:)

From what I've seen, I think people are trying to make Adam work, rather than accept it's actually not the best solution in general for most problems. As the blog post states, people were wowed by the idea it would make training twice as fast, but then there have been many indications that Adam (and similar methods) produce inferior solutions to plain SGD.

There are powerful proofs for the strength of SGD, and I'd take them over 'experimental' ('scientific method') contortions any day.

There is also a yet-to-be-implemented option for SGD in DL4J (and probably most other frameworks) that can achieve significant speed ups that won't affect SGD's proven characteristics (schemes to back prop on examples in a batch with highest loss, at least for the early stages of the training).

**Comment by stolsvik** [comment]
I thought that SGD had known problems, e.g. in "ravines" and "saddle points" - so to me it seems like there are ample opportunities for improvements wrt. to optimizers? And that these dynamic algorithms definitely have merit - at least momentum and nesterovs are obvious improvements?

**Comment by treo** [comment]
If I'm reading this correctly, DL4J actually already implemented it correctly in the first place:
https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/updater/UpdaterBlock.java#L145-L161

**Comment by stolsvik** [comment]
Based on my absolutely minimal understanding of the dl4j innards and looking at that code, it certainly seems so!

However, the only thing to maybe wonder about, is that the first article I linked to (the fast.ai) mentions the weight decay step as 

> when applying vanilla SGD it’s equivalent to updating the weight like this:
> w = w - lr * w.grad - lr * wd * w

And then further with e.g. momemtum:
> moving_avg = alpha * moving_avg + (1-alpha) * w.grad 
> w = w - lr * moving_avg - lr * wd * w

Notice how the weight decay parameter is multiplied with the learning rate.

This seems to not be the case in the dl4j code?

**Comment by stolsvik** [comment]
One more thing: The "postApply" seems to actually run on the already gradient-updated weights (i.e. the "Wn+1'"), or do I misunderstand? In the quoted functions, the weight decay is applied based on the existing weights ("Wn").

This is probably not extremely significant, but it departs slightly more from the "l2" logic, where the weights are decayed "inside" the function, and thus uses the "Wn" weights, not the gradient-updated "Wn+1" weights.

**Comment by stolsvik** [comment]
First: I then misunderstood one aspect: The postApply adds to the _update_ (which will be applied to the weights), not the weights (which would already have been updated by the .. update). So the weight decay applies to x, not x'. Okay.

However, I feel it makes some (intuitive?) sense that the learning rate should be taken into consideration when decaying the weights? If you have a high learning rate, then the weight decay should be heavier than if the learning rate is small? Which - with a static learning rate - would make no difference, just find an appropriate decay. But if you start to change the learning rate (as one apparently should do, judging by everything I read these days!), that is, use a schedule on it, typically going from high to low, it could make a difference? One might start out with appropriate decay based on a high learning rate, but then, as the learning rate lowers, the decay becomes "overpowering" compared to the updates?

**Comment by stolsvik** [comment]
According to this paper https://arxiv.org/abs/1705.08292 _"The Marginal Value of Adaptive Gradient Methods in Machine Learning"_, @tom-adsfund is right. One should apparently just ditch adaptive optimizers entirely, and go for SGD (with some added momentum), at least if it is the end result (generalization to new input) that is of interest..!

**Comment by treo** [comment]
I'm not against exploring other options. But since we already have adaptive updaters there is also no reason to ditch them, after all they do work pretty well in dl4j. 

If you want to schedule a different weight decay, all you have to do is change the l2 parameter during training.

**Comment by loshchil** [comment]
@treo @stolsvik 
> Our experiments show that weight decay is not like learning rates or momentum and the best value
> should remain constant through the training (i.e., cyclical weight decay is not useful).

Because they use the common way where L2-based gradients and loss-based gradients are coupled. Our paper says:

> It should be noted that when L2 regularization is used, weight decay contributes to the batch gradient and thus effectively is scheduled in the same way as the learning rate

Consider the common way to use SGD with L2 where 
```
update = gradient + w*x
x = x - alpha * update = x - alpha*gradient - alpha*w*x
```
Note that when you schedule `alpha` in the original SGD with L2, you effectively schedule weight decay.
Therefore, when we decouple the two, we use eta to set a schedule
`x' = x - eta*alpha*gradient - eta*w*x`
where `alpha `is the initial learning rate used primarily as a scaling factor.

There might be a problem with you current implementation. 
If I understand correctly, in your implementation you have 
`x' = x - alpha*gradient - w*x`
so your `w*x` is not multiplied by `alpha` nor by `eta`. It is fine only if you don't plan to schedule the two, but in almost all practical situations learning rate of SGD is scheduled. Therefore, by not doing it your  weight decay becomes relatively large over time, e.g., 10 times relatively larger once `alpha` is reduced by a factor of 10.  As a result, you have to select your `w` very conservatively to avoid divergence and this conservatism effectively disables regularization opportunities you can get from weight decay/L2. As a side effect, it might be difficult to reproduce SOTA results (without any loss of performance) for image classification datasets which heavily rely on L2/weight decay regularization. Note that by multiplying your `w*x` by `alpha`, you will couple  `w` and `alpha` again and have some problems of hyperparameter selection that we discussed in the paper. You can consider to use `eta` if it does not break your existing codebase.
I am not familiar with your framework, so I am sorry if everything is correct there and what I am saying does not apply.

**Comment by tom-adsfund** [comment]
@treo @stolsvik 
The holy grail of Adam speed, SGD generalization, and theoretical and empirical guarantees seems to have been found in "Padam" (University of Virginia):

https://arxiv.org/pdf/1806.06763.pdf

"Can we take the best from both Adam and SGD, i.e., design an algorithm that not only enjoys
the fast convergence rate as Adam, but also generalizes as well as SGD? In this paper, we answer this question affirmatively... We show that the Padam’s performance is also theoretically guaranteed."

https://arxiv.org/pdf/1808.05671.pdf

@loshchil Thanks for your work! I had the idea of pursuing updates with highest loss independently, and was very happy to find you'd already researched it!

**Comment by loshchil** [comment]
@tom-adsfund It is amazing to see that today's suggested algorithm Padam shows that yesterday's suggested algorithm Amgsrad does not work as promised (see Figure 2 in https://arxiv.org/pdf/1806.06763.pdf) :)

**Comment by tom-adsfund** [comment]
@stolsvik sure (I assume you're kind of commenting on a 'flavor of the day' issue with research), which is why I only posted it after seeing the second paper I linked to, which confirmed that the convergence rate is provably as good as SGD.

What I think is more interesting is that those graphs in figure 2 show that unless you're trying to optimize, SGD with momentum is great, which is what I started the thread saying. And so unless "Padam" is easy to implement, I'd stick with SGD and optimize elsewhere.

**Comment by agibsonccc** [comment]
Closing — this issue is over 7 years old with no recent activity. The codebase has undergone a major rewrite. If this is still relevant, please open a new issue with current reproduction steps.
